openedx/core/djangoapps/content_staging/models.py

```diff
@@ -0,0 +1,126 @@
+"""
+Models for content staging (and clipboard)
+"""
+import logging
+
+from django.contrib.auth import get_user_model
+from django.core.exceptions import ValidationError
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+from opaque_keys.edx.django.models import UsageKeyField
+from opaque_keys.edx.keys import LearningContextKey
+
+from openedx.core.djangoapps.content.course_overviews.api import get_course_overview_or_none
+
+log = logging.getLogger(__name__)
+
+User = get_user_model()
+
+
+class StagedContent(models.Model):
+    """
+    Each StagedContent instance represents a "piece" of content (e.g. a single
+    XBlock, or a single Unit, or a single Subsection of a course) that is not
+    currently part of any course or library, but which is available to be copied
+    into a course or library.
+
+    Use as a clipboard: for any given user, the most recent row with
+    purpose=CLIPBOARD is the "current" clipboard content. But it can only be
+    pasted if its status is READY.
+    """
+
+    class Meta:
+        verbose_name_plural = _("Staged Content")
+
+    class Status(models.TextChoices):
+        """ The status of this staged content. """
+        # LOADING: We are actively (asynchronously) writing the OLX and related data into the staging area.
+        # It is not ready to be read.
+        LOADING = "loading", _("Loading")
+        # READY: The content is staged and ready to be read.
+        READY = "ready", _("Ready")
+        # The content has expired and this row can be deleted, along with any associated data.
+        EXPIRED = "expired", _("Expired")
+        # ERROR: The content could not be staged.
+        ERROR = "error", _("Error")
+
+    id = models.AutoField(primary_key=True)
+    # The user that created and owns this staged content. Only this user can read it.
+    user = models.ForeignKey(User, null=False, on_delete=models.CASCADE)
+    created = models.DateTimeField(null=False, auto_now_add=True)
+    # What this StagedContent is for (e.g. "clipboard" for clipboard)
+    purpose = models.CharField(max_length=64)
+    status = models.CharField(max_length=20, choices=Status.choices)
+
+    block_type = models.CharField(
+        max_length=100,
+        help_text=_("""
+            What type of content is staged. Only OLX content is supported, and
+            this field must be the same as the root tag of the OLX.
+            e.g. "video" if a video is staged, or "vertical" for a unit.
+        """),
+    )
+    olx = models.TextField(null=False, blank=False)
```

Comment:
Predicting future requirements is a risky business :p I can imagine it, sure, but I have no idea if we'll think it's a good idea at some point in the future or not. I can also imagine StagedContent being a mutable work area instead of a read-only staging area, but I'm also unsure if that will be a good idea or not.